### PR TITLE
RSS Feed loader

### PR DESCRIPTION
**Pull request by sidmohanty11**
## Description

This PR adds support for loading RSS feeds directly from Embedchain.

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

## How Has This Been Tested?

- [x] Test Script (please provide)

```python3
from embedchain import Pipeline as App

app = App.from_config("config.yaml")

# app.add("https://news.ycombinator.com/rss", data_type="rss_feed")
# app.add("https://keepingupwith.ai/feed", data_type="substack")
app.add("https://www.lennysnewsletter.com/feed", data_type="substack")

response = app.query("Who is Brian Chesky?")
```

Output,

```
Inserting batches from 0 to 100 in chromadb
Inserting batches from 100 to 130 in chromadb
Successfully saved https://www.lennysnewsletter.com/feed (DataType.SUBSTACK). New chunks count: 130
Brian Chesky is the co-founder and CEO of Airbnb.
```

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

File: embedchain/loaders/rss_feed.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+
+@register_deserializable
+class RSSFeedLoader(BaseLoader):
+    """Loader for RSS Feed."""
+
+    def load_data(self, url):
+        """Load data from a rss feed."""
+        output = self.get_rss_content(url)
+        doc_id = hashlib.sha256((str(output) + url).encode()).hexdigest()
```
Comment (deshraj) [inline embedchain/loaders/rss_feed.py:20]: Should doc_id only contain the URL so that we know if the content changed for the hash or not?

Comment (deven298) [reply embedchain/loaders/rss_feed.py:20]: This is the right way to know if the content changed or not. If we keep only the url, we will not be able to know the delta.

File: embedchain/loaders/rss_feed.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+        return metadata
+
+    @staticmethod
+    def get_rss_content(url: str):
+        output = []
+        loader = LangchainRSSFeedLoader(urls=[url])
+        data = loader.load()
+
+        for entry in data:
```
Comment (deshraj) [inline embedchain/loaders/rss_feed.py:40]: This is going to be really slow if the RSS feed is quite long. You might want to use multithreading to speed up the loader.

File: embedchain/utils.py [hunk 3]
```diff
@@ -149,6 +148,10 @@ def is_openapi_yaml(yaml_content):
             logging.debug(f"Source of `{formatted_source}` detected as `notion`.")
             return DataType.NOTION
 
+        if url.netloc in {"substack.com"}:
+            logging.debug(f"Source of `{formatted_source}` detected as `substack`.")
+            return DataType.SUBSTACK
```
Comment (deshraj) [inline embedchain/utils.py:153]: What if someone has a custom domain name for substack url? Do they have to pass the `data_type=substack`? 

Also, if I am trying to just load a particular substack post as a webpage, that will be loaded as substack data type, right?

File: embedchain/loaders/rss_feed.py (hunk 2 above)
Comment (sidmohanty11) [reply embedchain/loaders/rss_feed.py:40]: Got it will update this

File: embedchain/utils.py (hunk 3 above)
Comment (sidmohanty11) [reply embedchain/utils.py:153]: Yeah! you're correct, should these be removed from auto-detecting types?

File: embedchain/loaders/rss_feed.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+            output.append(
```
[the remark this answers is not in the document]
Comment (sidmohanty11) [reply embedchain/loaders/rss_feed.py:41]: Thanks a lot! Will update this

File: embedchain/loaders/rss_feed.py [hunk 5]
```diff
@@ -0,0 +1,49 @@
+import hashlib
+
+try:
```
[the remark this answers is not in the document]
Comment (sidmohanty11) [reply embedchain/loaders/rss_feed.py:3]: Got it!

**Comment by sidmohanty11** [comment]
> @sidmohanty11 Seems like substack loader is calling rss loader in its load function. Any particular reason why we need both? Am I missing something?

You're 100% correct, as of now this is incomplete. We need to figure out a way to get all RSS feed posts for the substack URL. Then we can map it through from there. Else scrape the remaining posts and add it here

File: embedchain/chunkers/rss_feed.py [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+from embedchain.helper.json_serializable import register_deserializable
+
+
+@register_deserializable
+class RSSFeedChunker(BaseChunker):
+    """Chunker for RSS Feed."""
+
+    def __init__(self, config: Optional[ChunkerConfig] = None):
+        if config is None:
+            config = ChunkerConfig(chunk_size=1000, chunk_overlap=0, length_function=len)
```
Comment (deshraj) [inline embedchain/chunkers/rss_feed.py:16]:
```suggestion
            config = ChunkerConfig(chunk_size=2000, chunk_overlap=0, length_function=len)
```

**Review by deshraj — APPROVED** [review]

**Pull request merged**